### PR TITLE
Context propagation requirements for messaging semantic conventions

### PR DESCRIPTION
**Pull request by pyohannes**
This PR is a result of discussions in the messaging workgroup and has been split off #192 for easier review.

The [existing messaging semantic conventions for tracing](https://github.com/open-telemetry/opentelemetry-specification/blob/a1a8676a43dce6a4e447f65518aef8e98784306c/specification/trace/semantic_conventions/messaging.md) implicitly impose certain requirements on context propagation mechanisms used: while context is usually propagated on a per-request basis, the messaging semantic conventions assume that context is propagated on a per-message basis.

This OTEP proposes an addition to the existing messaging semantic conventions to make this requirement explicit. This is very important for reasoning about design questions regarding trace and span structure. Furthermore, explicating those requirements is of crucial importance for coming up with standardized context propagation mechanisms for messaging protocols. Some drafts for such standardization exist ([AMQP](https://w3c.github.io/trace-context-amqp/), [MQTT](https://w3c.github.io/trace-context-mqtt/), [CloudEvents](https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md)), and the changes in this PR should help to verify those drafts against the requirements of OpenTelemetry semantic conventions.

The addition proposed in this document does not break or invalidate any of the existing semantic conventions for messaging, but rather makes an implicit requirement explicit.

### CODE REVIEW

**Review by lmolkova — APPROVED** [review]

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (yurishkuro) [inline text/trace/0205-messaging-semantic-conventions-context-propagation.md:81]: > originate from different traces

Not clear why this is. What is the different "trace" on the consumer side? What workflow does that trace represent?

Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:81]: The trace on the consumer side represents the workflow of fetching a message from the intermediary and possibly processing it.

The consumer sends an HTTP GET request to the intermediary to receive the message. This request will propagate context from the consumer to the intermediary, but it doesn't propagate any context from the intermediary (or producer) to the consumer. So, via plain HTTP context propagation, the consumer has no way of obtaining context information from the intermediary (or producer), and thus cannot create parent/child or link relationships with the producer trace.

Comment (yurishkuro) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:81]: It can certainly be treated as a trace (sans the processing of the message, that makes less sense), but as you describe later those are not the same category of traces, you call it "transport" trace.

So I think this whole use case could be made much clearer by stating that the problem is with potentially overlapping workflows where more than one trace could be active during specific operations. I wanted to say "during the same tracepoints", but that does not seem to be the case, because each of the concurrent traces uses their own tracepoints, you explicitly mention that saying that message broker HTTP exchanges do not need to be recorded in the logical message processing call graph.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (joaopgrassi) [inline text/trace/0205-messaging-semantic-conventions-context-propagation.md:156]:
```suggestion
Currently, instrumentations have to decide how to attach and extract context from
```

Not sure if we have `instrumentor` in our glossary or if it's a known term in OTel? Maybe we should check if this is OK to use, to avoid confusion?

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (joaopgrassi) [inline text/trace/0205-messaging-semantic-conventions-context-propagation.md:160]:
```suggestion
propagate context using well-established messaging protocols.
```
Maybe? I feel `popular` is a loaded word that can mean different things for different people :)

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (joaopgrassi) [inline text/trace/0205-messaging-semantic-conventions-context-propagation.md:162]:
```suggestion
There are several working-in-progress efforts to standardize context propagation for different
```
Maybe?

Comment (joaopgrassi) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:162]: This way, we could also drop the  `Those standards are in draft states and/or are not widely adopted yet` part below, since it's implicit that if they are "work-in-progress" that they are drafts and not adopted.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (joaopgrassi) [inline text/trace/0205-messaging-semantic-conventions-context-propagation.md:170]: This part is a bit confusing to me. I think it's trying to say 2 things in the same paragraph. 

>  It is planned to drive those standards to a stable state and to make sure they cover requirements put forth by these semantic conventions

Who is exactly involved in "planned to drive those standards"? Is it us - OTel?  If so, then maybe making it more clear like:

> OpenTelemetry is going to help drive these standards, making sure they are compatible and cover the requirements put forth by these semantic conventions

And: 

> Finally, these semantic conventions should give a clear and stable recommendation for each protocol and scenario

I think this is a bit redundant? The paragraph above `in the future these conventions should list recommended ways..` says the same thing. I guess we can still keep this as a more "strong" assertion, but rephrase it a bit and make it a separated statement from the above:

> Once the standards reach a stable state, these semantic conventions will give a clear and stable recommendation for each protocol and scenario.

WDYT?

>

**Review by joaopgrassi — APPROVED** [review]
I think this overall makes sense. I left a few minor things, mostly around wording and structure of text. Feel free to address them or not.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:156]: The glossary contains the term "instrumentation author". I'll use that.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:133]: Some people in the workgroup expressed a preference for "creation context" (as opposed to "application context"). I don't have a strong opinion on that, if people think "application context" is clearer, we can use that name.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:83]: I would need to be forwarded _to_ the consumer, but it would need to be forwarded _by_ the HTTP GET request, when only using HTTP context propagation (without messaging specific context propagation).

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md]: You are right. I removed the reference to the example.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md]: I split the sentence up to make clear this refers to individual examples.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md]: I removed the must. The strongest we want to use is SHOULD.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:170]: > Who is exactly involved in "planned to drive those standards"? Is it us - OTel?

No, those standards are outside OTel. However, people working on those are counting on OTel to validate whether the specification the came up with makes sense and can be stable.

I agree this paragraph can be confusing. I will simplify it and just go with the last suggestion you made:

> Once the standards reach a stable state, these semantic conventions will give a clear and stable recommendation for each protocol and scenario.

Comment (lmolkova) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md]: BTW, I know we're not defining the standards, and having them defined would take a lot of time. Can we still have some soft recommendations of what people should do in absence of those?

E.g. "while the standards are being defined, we suggest that instrumentations that follow this convention use [Propagator API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#propagators-api) to propagate context using default or user-configured propagator."

Maybe we can also discuss it at the Thu meeting - I think we need to understand how it can evolve after AMQP/MQTT and other standards come into play and new instrumentations will have to stay backward-compatible.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:82]: > So I think this whole use case could be made much clearer by stating that the problem is with potentially overlapping workflows where more than one trace could be active during specific operations.

The problem is with providing the context information from the producer to the consumer, so the consumer can create link or parent/child relationships with the consumer.

I will try to improve the example so this becomes clearer.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (joaopgrassi) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:133]: We could try getting a consensus during the next meeting maybe? I think this is an important concept and a clear and easy to understand name could go a long way in helping folks grasp what we are suggesting.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md]: @lmolkova We can discuss this on Thursday.

From my point of view, this PR makes requirements for context propagation clear, but it doesn't give any recommendations for context propagation on top of that. As at this point we cannot give any stable recommendations (as currently none of the standard messaging protocols has a stable specification for context propagation established), I think we should refrain from giving any recommendations.

Also, it would be hard to give generic recommendations, as those recommendations would always be specific to certain protocols and messaging systems.

However, I'm open to other suggestions here, I'll put it on the agenda for Thursday.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:138]: We will call this _message context layer_ for now.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md]: Summarizing the discussion in today's messaging workgroup. Giving clear recommendations for context propagation requires 3 sequential steps:
1. Making requirements for context propagation clear (this PR).
2. Verifying that protocols (AMQP, MQTT, CloudEvents) satisfy those requirements, if so, drive context propagation extensions for those protocols to a stable state.
3. Recommend stable context propagation mechanisms for those protocols in OpenTelemetry semantic conventions.

Before all three steps are done, we will not really be able to give stable recommendations for context propagation, and users will be left with a certain (intended) uncertainty here.

Comment (lmolkova) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md]: Created https://github.com/w3c/trace-context-binary/issues/19 on the W3C binary protocol to prevent potential header name collisions between formats

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (blumamir) [inline text/trace/0205-messaging-semantic-conventions-context-propagation.md:126]: I would also mention that attaching a context per message is not always possible technically (redis pub sub, socket.io etc), and in those cases, there will be no correlation between producer and consumer. Although the SHOULD  word covers this case, it is so common that I think it deserves to be explicitly mentioned, and also communicate to readers that it is expected to happen and the implications.

**Review by blumamir — APPROVED** [review]
LGTM
Thank you Johannes

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:128]: I added a remark about this in the section above.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (dpauls) [inline text/trace/0205-messaging-semantic-conventions-context-propagation.md:152]: I know this is a "future", but I think intermediaries that are not instrumented might pass through the transport context unaltered. 

If we consider a message following a path:
producer --> messaging SDK --> intermediary --> messaging SDK --> consumer

If the SDK on the producer side includes a transport context and the intermediary isn't instrumented, having the transport context available to the consumer SDK could be useful. The SDK may instrument details like "received from intermediary", "enqueued to application", "delivered to application". I think these should be considered part of the transport trace, even if the intermediary didn't generate spans and modify the context.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (dpauls) [inline text/trace/0205-messaging-semantic-conventions-context-propagation.md]: Specifically, is this saying that they must define how both creation context and transport context will be carried as spearate entities? I think this should be a condition and stated a bit more clearly here.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (dpauls) [inline text/trace/0205-messaging-semantic-conventions-context-propagation.md]: I think last Thursday a suggestion was made to use the term "message context" instead of "creation context". Is this still being considered?

**Review by dpauls — CHANGES_REQUESTED** [review]

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (dpauls) [inline text/trace/0205-messaging-semantic-conventions-context-propagation.md:177]: I think we want the standards to reach a stable state **and** define how both message creation context and transport context should be carried in the message. The goal is to provide direction here what we would like the standards to achieve before being adopted by these semantic conventions.

Here's an example of what I think could go wrong without specific direction. The current MQTT draft standard defines a single context only. One path forward could be for the MQTT standard to declare that "this context is the message creation context and must not be modified by intermediaries" in the draft. This would allow it to be adopted here, but this would leave a gap for transport context propagation. 

Even if these conventions don't define transport layer tracing for messaging now, we want it to evolve to include transport-layer tracing. So I think we should provide guidance as to what is generally needed by those standards for messaging.

```suggestion
Once the standards reach a stable state and define how the message creation
context and transport context are represented, these semantic conventions will
give a clear and stable recommendation for each protocol and scenario.
```

**Review by dpauls — CHANGES_REQUESTED** [review]

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (joaopgrassi) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:152]: Not sure if I follow but if the intermediary is not instrumented, how would the transport trace continue? Somehow there must be some code that reads the incoming remote span and either creates a new child span or simply forwarded the same (which I'm not it's a good approach).

Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:152]: I think the main point we wanted to make here was, that we cannot make any strong assumptions regarding the transport context. By definition, intermediaries can change it. So, possibly, intermediaries might alter it, keep it as it is (that's the scenario @dpauls points out), or drop it. The main point we wanted to make is, that in contrast to the message context, the consumer cannot rely on receiving a transport context, even if the producer sent one.

I can also remove the sentence in question. I think at this point we don't want to make any assumptions or requirements about what un-instrumented intermediaries would or should do.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:177]: That makes sense to me. If nobody objects, I will accept this change.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md]: I'll resolve this, because it's covered by this discussion: https://github.com/open-telemetry/oteps/pull/205#discussion_r898403880

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (dpauls) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:152]: > Not sure if I follow but if the intermediary is not instrumented, how would the transport trace continue? 
If an intermediary doesn't even understand the concept of tracing, if it sees a message with a `traceparent` header or property, it is probably going to treat it the same as it would a `foo` or a `bar` header/property. Different intermediaries might do different things here, but a reasonable choice is to just propagate those through unchanged towards the consumer.

I agree with all three possibilities (alter, propagate, drop) that @pyohannes mentioned. While this last sentence uses a "might" which is in no way prescriptive, I think it makes sense to either mention all possibilities, or also removing the sentence would be reasonable as well. Singling out the "drop" scenario might cause some readers to interpret this as being a preferable behaviour.

Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md]: I removed the last sentence of this paragraph.

**Review by dpauls — APPROVED** [review]

**Review by yurishkuro — APPROVED** [review]

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (carlosalberto) [inline text/trace/0205-messaging-semantic-conventions-context-propagation.md:86]: `producer to intermediary` instead?

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (carlosalberto) [inline text/trace/0205-messaging-semantic-conventions-context-propagation.md:97]: When adding this to the Specification, concise examples (say, "this is how Kafka supports this") will be really nice to have.

File: text/trace/0205-messaging-semantic-conventions-context-propagation.md
Comment (pyohannes) [reply text/trace/0205-messaging-semantic-conventions-context-propagation.md:86]: This describes a pull-scenario, and the HTTP request referred to is the one by which the consumer fetches a message from the intermediary, so "consumer to the intermediary" is correct. 

> [...] _the consumer_ sends an HTTP GET request to the intermediary to fetch a message, the message is returned as part of the response.

**Comment by carlosalberto** [comment]
@open-telemetry/specs-approvers Please review. As previously mentioned, the messaging working group approved this OTEP already, we only need one more Spec approval.

**Review by jmacd — APPROVED** [review]

**Pull request merged**
